clientapi/routing/auth_fallback.go

```diff
@@ -0,0 +1,179 @@
+            data-sitekey="{{.siteKey}}"
+            data-callback="captchaDone">
+        </div>
+        <noscript>
+        <input type="submit" value="All Done" />
+        </noscript>
+        </div>
+    </div>
+</form>
+</body>
+</html>
+`
+
+// successTemplate is an HTML template presented to the user after successful
+// recaptcha completion
+const successTemplate = `
+<html>
+<head>
+<title>Success!</title>
+<meta name='viewport' content='width=device-width, initial-scale=1,
+    user-scalable=no, minimum-scale=1.0, maximum-scale=1.0'>
+<link rel="stylesheet" href="/_matrix/static/client/register/style.css">
+<script>
+if (window.onAuthDone) {
+    window.onAuthDone();
+} else if (window.opener && window.opener.postMessage) {
+    window.opener.postMessage("authDone", "*");
+}
+</script>
+</head>
+<body>
+    <div>
+        <p>Thank you!</p>
+        <p>You may now close this window and return to the application.</p>
+    </div>
+</body>
+</html>
+`
+
+// serveTemplate fills template data and serves it using http.ResponseWriter
+func serveTemplate(w http.ResponseWriter, templateHTML string, data map[string]string) {
+	t := template.Must(template.New("response").Parse(templateHTML))
+	if err := t.Execute(w, data); err != nil {
+		panic(err)
+	}
+}
+
+// AuthFallback implements GET on /auth/{authType}/fallback/web?session={sessionID}
+func AuthFallback(
+	w http.ResponseWriter, req *http.Request, authType string,
+	cfg config.Dendrite,
+) *util.JSONResponse {
+	sessionID := req.URL.Query().Get("session")
+
+	if sessionID == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		_, err := w.Write([]byte("Session ID not provided"))
+		if err != nil {
+			res := httputil.LogThenError(req, err)
+			return &res
+		}
+		return nil
+	}
+
+	serveRecaptcha := func() {
+		data := map[string]string{
+			"myUrl":   req.URL.String(),
+			"session": sessionID,
+			"siteKey": cfg.Matrix.RecaptchaPublicKey,
+		}
+		serveTemplate(w, recaptchaTemplate, data)
+	}
+
+	serveSuccess := func() {
+		data := map[string]string{}
+		serveTemplate(w, successTemplate, data)
+	}
+
+	if req.Method == http.MethodGet {
+		// Handle Recaptcha
+		if authType == authtypes.LoginTypeRecaptcha {
+			if cfg.Matrix.RecaptchaPublicKey == "" {
+				w.WriteHeader(http.StatusInternalServerError)
+				_, err := w.Write([]byte("This Homeserver doesn't have a recaptcha public key"))
+				if err != nil {
+					res := httputil.LogThenError(req, err)
+					return &res
+				}
+				return nil
+			}
+
+			serveRecaptcha()
+			return nil
+		}
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("Unknown auth stage type"),
+		}
+	} else if req.Method == http.MethodPost {
+		clientIP := req.RemoteAddr
+		err := req.ParseForm()
+		if err != nil {
+			res := httputil.LogThenError(req, err)
+			return &res
+		}
+
+		response := req.Form.Get("g-recaptcha-response")
+		if err := validateRecaptcha(&cfg, response, clientIP); err != nil {
+			serveRecaptcha()
```

Comment:
Why don't we pass the error from `validateRecaptcha` to the client?

Reply:
Do you mean just sending the error JSON response or showing a page like "An error has occured"?

Reply:
I think we should just pass the error returned by `validateRecaptcha` here since `err` is already a `JSONResponse`. However, we'll probably also want to refine the error handling in `validateRecaptcha` to hide some unnecessary details from the clients and to log them in the backend.

Reply:
Now I log and pass the error response to client.